micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java

```diff
@@ -0,0 +1,279 @@
+            boundMeters.forEach((metricName, meters) -> {
+                if (meters.size() > 1) {
+                    //Find largest number of tags
+                    int maxTagsSize = 0;
+                    for (Meter meter : meters) {
+                        int size = meter.getId().getTags().size();
+                        if (maxTagsSize < size) maxTagsSize = size;
+                    }
+                    //Remove meters with lower number of tags
+                    for (Meter meter : meters) {
+                        if (meter.getId().getTags().size() < maxTagsSize) registry.remove(meter);
+                    }
+                }
+            });
+        }
+    }
+
+    @NotNull private Meter bindMeter(MeterRegistry registry, Metric metric) {
+        String metricName = metricName(metric);
+        Meter meter;
+        if (metricName.endsWith("total") || metricName.endsWith("count")) {
+            meter = registerCounter(registry, metric, metricName, extraTags);
+        } else if (metricName.endsWith("min")
+                || metricName.endsWith("max")
+                || metricName.endsWith("avg")) {
+            meter = registerGauge(registry, metric, metricName, extraTags);
+        } else if (metricName.endsWith("rate")) {
+            meter = registerTimeGauge(registry, metric, metricName, extraTags);
+        } else {
+            meter = registerGauge(registry, metric, metricName, extraTags);
+        }
+        return meter;
+    }
+
+    private TimeGauge registerTimeGauge(MeterRegistry registry, Metric metric, String metricName, Iterable<Tag> extraTags) {
+        return TimeGauge.builder(metricName, metric, TimeUnit.SECONDS, toMetricValue(registry))
+            .tags(metricTags(metric))
+            .tags(extraTags)
+            .description(metric.metricName().description())
+            .register(registry);
+    }
+
+    private Gauge registerGauge(MeterRegistry registry, Metric metric, String metricName, Iterable<Tag> extraTags) {
+        return Gauge.builder(metricName, metric, toMetricValue(registry))
+            .tags(metricTags(metric))
+            .tags(extraTags)
+            .description(metric.metricName().description())
+            .register(registry);
+    }
+
+    private FunctionCounter registerCounter(MeterRegistry registry, Metric metric, String metricName, Iterable<Tag> extraTags) {
+        return FunctionCounter.builder(metricName, metric, toMetricValue(registry))
+            .tags(metricTags(metric))
+            .tags(extraTags)
+            .description(metric.metricName().description())
+            .register(registry);
+    }
+
+    private ToDoubleFunction<Metric> toMetricValue(MeterRegistry registry) {
+        return metric -> {
+            //Double-check if new metrics are registered; if not (common scenario)
+            //it only adds metrics count validation
+            checkAndBindMetrics(registry);
+            if (metric.metricValue() instanceof Double) {
```

Comment:
Do we know what other kinds of values are expected from Kafka metrics? If it were say a String, us returning `0.0` could be confusing. I wonder if `NaN` is better when the value isn't a double.

Reply:
I was considering `NaN` before reading this <https://www.robustperception.io/get-thee-to-a-nannary> but it will depend on how Micrometer semantics work on this case.

After taking another look at Kafka source-code I found there are certain metrics that return generic values, where we can find integers. I'll add this case to this function.

Reply:
The advice in there seems to be do not register a metric if it does not have a numeric value. That seems like good advice. Let's not register metrics with non-numeric values. Then, assuming the value type cannot change for a given metric, we don't need to handle the non-numeric case in the `ToDoubleFunction`